lemur/plugins/lemur_gcp/certificates.py

```diff
@@ -0,0 +1,81 @@
+from flask import current_app
+from google.cloud.compute_v1.services import ssl_certificates
+
+from lemur.common.utils import parse_certificate, split_pem
+from cryptography.hazmat.primitives import hashes
+
+
+def get_name(body):
+    """
+    We need to change the name of the certificate that we are uploading to comply with GCP naming standards.
+    The cert name will follow the convention "ssl-{SHA1 fingerprint}". This is guaranteed to be unique
+    across CAs and complies with naming restrictions from the GCP API.
+    """
+    cert = parse_certificate(body)
+    # SHA1 is 160 bits => 40 hex digits
+    fingerprint = cert.fingerprint(hashes.SHA1()).hex() # nosec B303
+    cert_name = f"ssl-{fingerprint}".lower()
+    # This should never happen
+    if len(cert_name) > 63:
+        raise Exception(f"Could not create certificate name for certificate: {body}")
+    return cert_name
+
+
+def full_ca(body, cert_chain):
+    # in GCP you need to assemble the cert body and the cert chain in the same parameter
+    return f"{body}\n{cert_chain}"
+
+
+def insert_certificate(project_id, ssl_certificate_body, credentials):
+    return ssl_certificates.SslCertificatesClient(credentials=credentials).insert(
+        project=project_id, ssl_certificate_resource=ssl_certificate_body
+    )
+
+
+def fetch_all(project_id, credentials):
+    client = ssl_certificates.SslCertificatesClient(credentials=credentials)
+    certs = []
+    for cert_meta in client.list(project=project_id):
+        try:
+            if cert_meta.type_ != "SELF_MANAGED":
+                continue
+            cert = parse_certificate_meta(cert_meta)
+            if cert:
+                certs.append(cert)
+        except Exception as e:
+            current_app.logger.error(
+                f"Issue with fetching certificate {cert_meta.name} from GCP. Action failed with the following "
+                f"log: {e}",
+                exc_info=True,
+            )
+            raise e
+    return certs
+
+
+def fetch_by_name(project_id, credentials, certificate_name):
+    client = ssl_certificates.SslCertificatesClient(credentials=credentials)
+    cert_meta = client.get(project=project_id, ssl_certificate=certificate_name)
```

Comment:
Will this method work as expected? Correct me if I'm wrong, but It looks like `GCPSourcePlugin.get_certificate_by_name` is passing in the lemur `certificate_name`. I think we need to lookup the certificate via the special GCP name we give a certificate.

Reply:
This is invoked immediately after `get_endpoints` is called (which returns endpoints with the attached cert by its API name). Check out [service.py](https://github.com/DataDog/lemur/blob/master/lemur/sources/service.py#L109) and [endpoints.py](https://github.com/DataDog/lemur/blob/75d0655d31ac8d393802a027cbb560c2186ac99c/lemur/plugins/lemur_gcp/endpoints.py#L58). The `certificate_name` here is the same as its API name